GNSS RTK: Add basic integration

ubieda opened the pull request:
### Description

This PR introduces RTK (Real Time Kinematics) to GNSS modems by providing a path to feed RTK correction data (proposed on #90363).

Detailed changes:
- Add APIs so RTK clients can be added (basic Serial client is added to demo).
- Add APIs to allow RTK clients to publish correction data, and for GNSS drivers to feed such data to their modems (applied to UBX F9P driver).
- Add RTCM3 protocol decoder (framing detection, validated with unit-test suite).
- Add GNSS RTK sample with python script to convert a PC in a Base-Station (requires a F9P module USB-connected), which feeds data to a Zephyr-powered board (over Serial port).

> [!NOTE]
> 1. This PR depends on F9P driver introduction: #90027
> 2. This PR will be submitted to present at the Architecture WG.
> Adding DNM until both conditions are resolved.

ubieda commented on subsys/rtk/protocol/rtcm3_crc.h:
```diff
@@ -0,0 +1,73 @@
+/*
```
I'd like some feedback WRT this file: As far as I'm concerned this CRC calculation is specific to RTCM3, however I don't know if it's preferred to put this under the more generic `include/zephyr/sys/crc.h`

ubieda replied:
Moved to `zephyr/sys/crc.h`

ubieda commented on subsys/rtk/Kconfig:
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2025 CogniPilot Foundation
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig GNSS_RTK
+	bool "GNSS RTK client"
```
select EXPERIMENTAL

ubieda commented on subsys/rtk/CMakeLists.txt:
Suggestion: Move to `drivers/gnss/*`

ubieda replied:
Moved to `subsys/gnss` instead.

ubieda reviewed (commented):
Comments captured from Arch WG review 250701

fabiobaltieri commented on drivers/gnss/gnss_u_blox_f9p.c:
```diff
@@ -516,6 +545,10 @@ static DEVICE_API(gnss, ublox_f9p_driver_api) = {
 												   \
 	static struct ubx_f9p_data ubx_f9p_data_##inst;						   \
 												   \
+	COND_CODE_1(CONFIG_GNSS_U_BLOX_F9P_RTK,							   \
```
`IF_ENABLED`

ubieda commented on samples/drivers/gnss/rtk/src/main.c:
```diff
@@ -0,0 +1,94 @@
+	if (data->info.fix_status != GNSS_FIX_STATUS_NO_FIX) {
+		if (gnss_get_latest_timepulse(dev, &timepulse) == 0) {
+			timepulse_ns = k_ticks_to_ns_near64(timepulse);
+			printf("Got a fix @ %lld ns\n", timepulse_ns);
```
@fabiobaltieri WRT this and the other comments on the sample code (main.c): this PR basically extends the existing in-tree sample https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/gnss/src/main.c.

I only put it separate to improve visibility. But if this will cause both to start differing, it may be more convenient to just add an `overlay-rtk.conf`and add the instructions on the existing readme.

ubieda committed:
crc: Add CRC24Q for RTCM3 checksum calculations

Used in GNSS RTK correction frames.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
tests: crc: Add test-case to validate CRC24Q RTCM3

Validating both a frame without CRC (should return calculated CRC), and
with CRC (should return 0 as a match).

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
gnss: rtk: Add basic integration

Incorporate the basic RTK API as well as a basic client integration
(serial) as a way to receive and publish the RTK data.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
tests: gnss: rtk: Add testsuite to validate RTCM3 protocol

To capture framing capabilities (only API so far).

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
modem: ubx: Change request buffer to be void

So this API can be used to send frames with a different encoding than
UBX. This enables UBX drivers to send RTCM3 correction frames using UBX
API, without having to switch over modem pipes.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
gnss: u-blox f9p: Add RTK integration to driver

Enable driver to consume RTK data-correction messages published
in order to enhance GNSS Navigation results.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
samples: gnss: Add ability to enable RTK

Expand this sample to also work with GNSS_RTK through a serial client.

Make this compatible to work with VMU_RT1170 which requires using the
F9P, instead of the M8 UBX module.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented on tests/subsys/modem/modem_ubx/src/main.c:
```diff
@@ -94,7 +94,7 @@ static inline void restore_ubx_script(void)
 		.script = {
 			.request = {
-				.buf = &test_req,
+				.buf = (const uint8_t *)&test_req,
```
You're right. Moved to its own commit now.

ubieda commented on samples/drivers/gnss/rtk/base_station_f9p.py:
```diff
@@ -0,0 +1,50 @@
+
+with (
+    serial.Serial("/dev/ttyACM0", 115200) as serialIn,
+    serial.Serial("/dev/ttyUSB0", 115200) as serialOut,
```
Sounds good. I've removed hardcoding both serial ports. Now one is detected by the VID:PID, the other one is a command-line argument (`--port`) for the board running the Zephyr app.

ubieda commented on samples/drivers/gnss/basic/README.rst:
```diff
@@ -1,5 +1,5 @@
-.. zephyr:code-sample:: gnss
-   :name: GNSS
+.. zephyr:code-sample:: gnss-basic
+   :name: GNSS Basic
```
ACK. Lumped sample-changes in a separate patch.

ubieda commented on lib/crc/crc24_sw.c:
Moved the API addition to its own patch, and added a crc-testcase in the following patch.

If possible, I'd like to keep it as part of this PR as this is where it's used (vs introducing an API with no in-tree users).

fabiobaltieri replied:
Sure, I think both approaches are valid.

ubieda commented on samples/drivers/gnss/rtk/src/main.c:
```diff
@@ -0,0 +1,94 @@
+	uint64_t timepulse_ns;
+	k_ticks_t timepulse;
+
+	if (data->info.fix_status != GNSS_FIX_STATUS_NO_FIX) {
```
Removed this diff by unifying both GNSS samples.

ubieda commented on samples/drivers/gnss/rtk/src/main.c:
```diff
@@ -0,0 +1,94 @@
+#define GNSS_SYSTEMS_PRINTF(define, supported, enabled)                                            \
+	printf("\t%20s: Supported: %3s Enabled: %3s\n",                                            \
+	       STRINGIFY(define), (supported & define) ? "Yes" : "No",                             \
+			 (enabled & define) ? "Yes" : "No");
```
Removed this diff by unifying both GNSS samples.

ubieda commented on subsys/gnss/rtk/rtk_publish.c:
```diff
@@ -0,0 +1,24 @@
+#include <zephyr/sys/iterable_sections.h>
+#include <zephyr/gnss/rtk/rtk.h>
+
+static K_SEM_DEFINE(semlock, 1, 1);
```
Made both changes.

fabiobaltieri commented on subsys/gnss/rtk/protocol/rtcm3.c:
```diff
@@ -0,0 +1,58 @@
+		uint16_t payload_len = RTCM3_FRAME_PAYLOAD_SZ(frame->hdr);
+		uint16_t remaining_bytes = buf_len - i;
+
+		if ((payload_len == 0) ||
+			((RTCM3_FRAME_SZ(payload_len)) > remaining_bytes)) {
```
nit: outer parenthesis here are redundant, no hard rule so fine if you want to keep them but I found it a bit silly, this is just hard to read

fabiobaltieri commented on subsys/gnss/rtk/serial/serial.c:
```diff
@@ -0,0 +1,96 @@
+
+		err = gnss_rtk_decoder_frame_get(work_buf + offset, len - offset,
+						 &frame, &frame_len);
+		if (err == 0) {
```
swap, if err continue

fabiobaltieri commented on subsys/gnss/rtk/serial/serial.c:
```diff
@@ -0,0 +1,96 @@
+
+			offset += frame_len;
+		}
+	} while ((err == 0) && (len > offset));
```
same as above, `} while (err == 0 && len > offset);`

fabiobaltieri commented on subsys/gnss/rtk/serial/serial.c:
```diff
@@ -0,0 +1,96 @@
+	(void)uart_irq_update(dev);
+
+	if (uart_irq_rx_ready(dev)) {
+		int ret = 0;
```
drop the initializer

fabiobaltieri commented on include/zephyr/modem/ubx.h:
```diff
@@ -53,7 +53,7 @@ struct modem_ubx_match {
 struct modem_ubx_script {
 	struct {
-		const struct ubx_frame *buf;
+		const uint8_t *buf;
```
if you do this then drop the casting in `subsys/modem/modem_ubx.c` `modem_ubx_run_script` since at that point it's not needed anymore,that said, since there's no point in the upstream code that accesses the data here as a byte array another option would be to change this to a `const void *` instead, that way you'd be able to drop all the castings in this PR and only keep the one in `modem_ubx_run_script`, and I don't think you'd lose anything really, the compiler would only make sure that the data pointed is constant and that's all you want.

bperseghetti reviewed (dismissed):
This is a really great contribution!

fabiobaltieri commented on include/zephyr/modem/ubx/protocol.h:
```diff
@@ -74,12 +74,19 @@ enum ubx_nav_fix_type {
 #define UBX_NAV_PVT_VALID_MAGN				BIT(3)
 
 #define UBX_NAV_PVT_FLAGS_GNSS_FIX_OK			BIT(0)
+#define UBX_NAV_PVT_FLAGS_GNSS_CARR_SOLN_FLOATING	(BIT(6))
+#define UBX_NAV_PVT_FLAGS_GNSS_CARR_SOLN_FIXED		(BIT(7))
```
drop the outer parenthesis, redundant, you only need to parenthesize arguments or operations, adding extra parenthesis randomly everywhere makes this a huge mess when you'll inevitably try to troubleshoot something when looking at the intermediate files

ubieda committed:
maintainers: Add GNSS Subsystem entries

To be covered by existing Drivers: GNSS, as per requested during the
review process. Added myself as a collaborator to follow-up with
RTK-related changes.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented on subsys/CMakeLists.txt:
```diff
@@ -16,6 +16,7 @@ add_subdirectory(canbus)
 add_subdirectory(debug)
 add_subdirectory(fb)
 add_subdirectory(fs)
+add_subdirectory(gnss)
```
Addressed. Added @bjarki-andreasen as maintainer to keep autonomy on GNSS-related decisions, listed myself as a collab. Would you like to be added there as well?

fabiobaltieri commented on subsys/gnss/rtk/rtk.c:
```diff
@@ -0,0 +1,24 @@
+#include <zephyr/sys/iterable_sections.h>
+#include <zephyr/gnss/rtk/rtk.h>
+
+void gnss_rtk_publish_data(const struct gnss_rtk_data *data)
+{
+	static K_SEM_DEFINE(publish_lock, 1, 1);
+
```
yeah not sure why it does not like it, it's certainly uncommon, that's probably why

ubieda commented on MAINTAINERS.yml:
```diff
@@ -2468,6 +2468,20 @@ Formatted Output:
+    - "area: GNSS"
+  tests:
+    - gnss
+
```
Done. Didn't rename it as this section is lumped alongside all the other Drivers. I suggest this housekeeping is done in a follow-up PR.

ubieda commented on drivers/gnss/gnss_u_blox_f9p.c:
```diff
@@ -187,6 +203,19 @@ static int ubx_f9p_msg_payload_send(const struct device *dev, uint8_t class, uin
+	 * it or not depending on the RTCM3 message type and its alignment with what the
+	 * GNSS modem has observed.
+	 */
+	(void)ubx_f9p_msg_send(dev, (const void *)data->data, data->len, false);
```
I don't think we should cast this to an UBX frame, as it's an RTCM3 msg. See commit:
`modem: ubx: Change request buffer to be void`

fabiobaltieri commented on MAINTAINERS.yml:
```diff
@@ -1569,21 +1569,26 @@ Release Notes:
     - include/zephyr/dt-bindings/gnss/
     - tests/drivers/build_all/gnss/
     - tests/drivers/gnss/
+    - tests/gnss/
```
```suggestion
    - tests/subsys/gnss/
```

ubieda replied:
Yep I realize I missed the folder. Pushed now fix.

bjarki-andreasen reviewed (dismissed):
Nice work!

kartben commented on subsys/gnss/rtk/serial/serial.c:
```diff
@@ -0,0 +1,99 @@
+		err = gnss_rtk_decoder_frame_get(work_buf + offset, len - offset,
+						 &frame, &frame_len);
+		if (err != 0) {
+			continue;
```
infinite loop?

ubieda replied:
Not sure I follow this: the loop break as soon as err != 0. 

Am I missing something here?

fabiobaltieri replied:
Oh I see, sorry what I meant was to indent the error handling and keep the "normal" case execution in line, it's a good coding practice and is used throughout the code base. Looking how this turned out (which I missed on my second pass) I think this should be reworked to avoid having the `err` evaluation split in two different points (the while and the continue branch), as written right now I think it's hard to follow for that reason.

kartben commented on include/zephyr/gnss/rtk/decoder.h:
```diff
@@ -0,0 +1,32 @@
+ * @return -EAGAIN if there's an incomplete frame starting at the data pointer.
+ * @return Other negative error code if decoding failed.
+ */
+int gnss_rtk_decoder_frame_get(uint8_t *buf, size_t buf_len,
```
```suggestion
int gnss_rtk_decoder_frame_get(const uint8_t *buf, size_t buf_len,
```

ubieda replied:
This was intentionally left as a mutable array because I was considering there may be some encoded frames with byte stuffing, hence leaving the option for the API to do things behind this API call.

It is not certainly the case with the only one implementation (RTCM3), but I'd prefer keep it this way unless you feel strongly about enforcing it until an implementation that requires it comes by.

kartben replied:
I guess I am just confused by the fact of having two functions named exactly the same but having [different signatures](https://github.com/zephyrproject-rtos/zephyr/pull/90890/files#diff-bb1190f6232750a33cab16130e770d0dd243fdea770932b3a753e016b5eebd4dR30-R31)

ubieda replied:
Should be consistent now. Removed stray header file under `subsys/gnss/rtk/decoder.h`

kartben commented on include/zephyr/gnss/rtk/decoder.h:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @return Zero if successful.
+ * @return -ENOENT if no frames have been decoded successfully.
+ * @return -EAGAIN if there's an incomplete frame starting at the data pointer.
```
you haven't implemented that error code tho?

ubieda replied:
Removed

kartben commented on subsys/gnss/rtk/protocol/rtcm3.c:
```diff
@@ -0,0 +1,58 @@
+	uint8_t payload[];
+} __packed;
+
+int gnss_rtk_decoder_frame_get(uint8_t *buf, size_t buf_len,
```
```suggestion
int gnss_rtk_decoder_frame_get(const uint8_t *buf, size_t buf_len,
```

ubieda replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/90890#discussion_r2217875470

kartben commented on subsys/gnss/rtk/protocol/rtcm3.c:
```diff
@@ -0,0 +1,58 @@
+int gnss_rtk_decoder_frame_get(uint8_t *buf, size_t buf_len,
+			       uint8_t **data, size_t *data_len)
+{
+	for (int i = 0 ; (i + RTCM3_FRAME_OVERHEAD - 1) < buf_len ; i++) {
```
```suggestion
	for (size_t i = 0 ; (i + RTCM3_FRAME_OVERHEAD - 1) < buf_len ; i++) {
```

kartben reviewed (changes requested):
Putting a request for change as I think at least the infinite loop concern should be addressed

fabiobaltieri commented on subsys/gnss/rtk/serial/serial.c:
```diff
@@ -0,0 +1,99 @@
+
+	do {
+		uint8_t *frame;
+		size_t frame_len = 0;
```
this does not need to be initialized

kartben commented on subsys/gnss/rtk/protocol/rtcm3.c:
```diff
@@ -0,0 +1,58 @@
+
+		if (crc24q_rtcm3((const uint8_t *)frame,
+				 RTCM3_FRAME_SZ(payload_len)) == 0) {
+			*data = (uint8_t *)frame;
```
coding guideline 70 / MISRA rule 11.8: "A cast shall not remove any const or volatile qualification from the type pointed to by a pointer"

kartben commented on subsys/gnss/rtk/serial/serial.c:
```diff
@@ -0,0 +1,99 @@
+		 * (e.g: 1 Hz), wait until all messages are received before
+		 * processing.
+		 */
+		k_work_reschedule(&gnss_rtk_process_work, K_MSEC(10));
```
return value must not be ignored (MISRA rule 17.7)

kartben commented on subsys/gnss/rtk/serial/serial.c:
```diff
@@ -0,0 +1,99 @@
+
+static void gnss_rtk_process_work_handler(struct k_work *work)
+{
+	static uint8_t work_buf[2048];
```
```suggestion
	ARG_UNUSED(work);

	static uint8_t work_buf[2048];
```

kartben commented on lib/crc/crc24_sw.c:
```diff
@@ -30,3 +33,62 @@ uint32_t crc24_pgp_update(uint32_t crc, const uint8_t *data, size_t len)
 	return crc;
 }
+
+/** This code is originally extracted from https://github.com/go-gnss/rtcm,
```
Interestingly this is a violation of coding guideline 24 ("The character sequences /* and // shall not be used within a comment"). That rule seems very impractical though :)

kartben commented on drivers/gnss/gnss_ubx_common.c:
```diff
@@ -27,19 +27,24 @@ void gnss_ubx_common_pvt_callback(struct modem_ubx *ubx, const struct ubx_frame
-			break;
-		default:
-			break;
 		}
```
MISRA rule 15.7 -- All if else if constructs shall be terminated with an else statement

kartben commented on drivers/gnss/gnss_ubx_common.c:
```diff
@@ -27,19 +27,24 @@ void gnss_ubx_common_pvt_callback(struct modem_ubx *ubx, const struct ubx_frame
-		switch (nav_pvt->fix_type) {
-		case UBX_NAV_FIX_TYPE_DR:
-		case UBX_NAV_FIX_TYPE_GNSS_DR_COMBINED:
+		if (nav_pvt->flags & UBX_NAV_PVT_FLAGS_GNSS_CARR_SOLN_FLOATING) {
```
MISRA rule 15.7 -- All if else if constructs shall be terminated with an else statement

fabiobaltieri commented on subsys/gnss/rtk/decoder.h:
is this file included from anywhere? `gnss_rtk_decoder_frame_get` is declared on the other header and the drivers include `zephyr/gnss/rtk/decoder.h`

ubieda replied:
Removed. This got leftover during directories refactoring while iterating.

ubieda commented:
@kartben @fabiobaltieri please revisit
